neo/Ledger/MemoryPool.cs

```diff
@@ -0,0 +1,165 @@
+            public PoolItem(Transaction tx)
+            {
+                Transaction = tx;
+                Timestamp = DateTime.UtcNow;
+            }
+        }
+
+        private readonly ConcurrentDictionary<UInt256, PoolItem> _mem_pool_fee = new ConcurrentDictionary<UInt256, PoolItem>();
+        private readonly ConcurrentDictionary<UInt256, PoolItem> _mem_pool_free = new ConcurrentDictionary<UInt256, PoolItem>();
+
+        public int Capacity { get; }
+        public int Count => _mem_pool_fee.Count + _mem_pool_free.Count;
+
+        public MemoryPool(int capacity)
+        {
+            Capacity = capacity;
+        }
+
+        public void Clear()
+        {
+            _mem_pool_free.Clear();
+            _mem_pool_fee.Clear();
+        }
+
+        public bool ContainsKey(UInt256 hash) => _mem_pool_free.ContainsKey(hash) || _mem_pool_fee.ContainsKey(hash);
+
+        public IEnumerator<Transaction> GetEnumerator()
+        {
+            return
+                _mem_pool_fee.Select(p => p.Value.Transaction)
+                .Concat(_mem_pool_free.Select(p => p.Value.Transaction))
+                .GetEnumerator();
+        }
+
+        IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
+
+        static void RemoveLowestFee(ConcurrentDictionary<UInt256, PoolItem> pool, int count)
+        {
+            if (count <= 0) return;
+            if (count >= pool.Count)
+            {
+                pool.Clear();
+            }
+            else
+            {
+                UInt256[] delete = pool.AsParallel()
+                    .OrderBy(p => p.Value.Transaction.NetworkFee / p.Value.Transaction.Size)
+                    .ThenBy(p => p.Value.Transaction.NetworkFee)
+                    .ThenBy(p => new BigInteger(p.Key.ToArray()))
+                    .Take(count)
+                    .Select(p => p.Key)
+                    .ToArray();
+
+                foreach (UInt256 hash in delete)
+                {
+                    pool.TryRemove(hash, out _);
+                }
+            }
+        }
+
+        static void RemoveOldest(ConcurrentDictionary<UInt256, PoolItem> pool, DateTime time)
+        {
+            UInt256[] hashes = pool
+                .Where(p => p.Value.Timestamp < time)
+                .Select(p => p.Key)
+                .ToArray();
+
+            foreach (UInt256 hash in hashes)
+            {
+                pool.TryRemove(hash, out _);
+            }
+        }
+
+        public bool TryAdd(UInt256 hash, Transaction tx)
+        {
+            ConcurrentDictionary<UInt256, PoolItem> pool;
+
+            if (tx.IsLowPriority)
+            {
+                pool = _mem_pool_free;
+            }
+            else
+            {
+                pool = _mem_pool_fee;
+            }
+
+            pool.TryAdd(hash, new PoolItem(tx));
+
+            if (Count > Capacity)
+            {
+                RemoveOldest(_mem_pool_free, DateTime.UtcNow.AddSeconds(-Blockchain.SecondsPerBlock * 20));
```

Comment:
This basically will always do nothing, because after each block is persisted it re-adds the transactions and they get restamped with a new timestamp. I'm working on some major adjustments to this code that will improve performance, fix priority inversion that happens when adding transactions back to the chain after persist, and preserve the timestamp of the transaction across blocks.